fix: Configuring ACL Plugin (#6233)
Kong/docs.konghq.com: app/_src/kubernetes-ingress-controller/guides/configure-acl-plugin.md

content_type: tutorial
---

Learn to configure the Kong ACL Plugin. To use the ACL Plugin you need at least one Authentication plugin. This example uses the JWT Auth Plugin.

{% include_cached /md/kic/prerequisites.md kong_version=page.kong_version disable_gateway_api=false%}

> {% include_cached /md/kic/class.md kong_version=page.kong_version %}

Reviewer: Do we need the `Create a configuration group` section? The `helm` install creates this resource automatically

Author: removed.

> {% include_cached /md/kic/http-test-routing.md kong_version=page.kong_version path='/lemon' name='lemon' %}

Author: I am not able to fix this. Because it is a nested include and we are using this in a lot of guides.

> {% include_cached /md/kic/http-test-routing.md kong_version=page.kong_version path='/lemon' name='lemon' %}

Reviewer: The indentation in the include was incorrect. See dfc3b35d672e7da9ea607c48119355e6d4e50c1c

After the first route is working, create a second pointing to the same Service:

{% include_cached /md/kic/http-test-routing-resource.md kong_version=include.kong_version path='/lime' name='lime' %}

## Add JWT authentication to the service

To add authentication in front of an API you just need to enable a plugin.
1. Create a KongPlugin resource.

    ```bash
    echo "
    apiVersion: configuration.konghq.com/v1
    kind: KongPlugin
    metadata:
      name: app-jwt
    plugin: jwt
    " | kubectl apply -f -
    ```
    The results should look like this:
    ```text
    kongplugin.configuration.konghq.com/app-jwt
    ```
1. Associate the plugin to the Ingress rules.

    ```bash
    kubectl annotate service echo konghq.com/plugins=app-jwt
    ```
    The results should look like this:
    ```text

>     service.networking.k8s.io/echo annotated

Reviewer: ```suggestion
    service/echo annotated
```

    ```

>     Any requests matching the proxying rules for `demo-get` and `demo` post now requires a valid JWT and the consumer for the JWT to be associate with the right ACL. Requests without credentials are rejected.

Reviewer: ```suggestion
    Any requests matching the proxying rules for `/lemon` and `/lime` now requires a valid JWT and the consumer for the JWT to be associate with the right ACL. Requests without credentials are rejected.
```

1. Send a request without the credentials.    

    ```bash
    curl -i -H 'Host:kong.example' $PROXY_IP/lemon
    ```
    The results should look like this:
    ```text
    HTTP/1.1 401 Unauthorized
    Date: Fri, 09 Dec 2022 23:51:35 GMT
    Content-Type: application/json; charset=utf-8
    Connection: keep-alive
    Content-Length: 26
    X-Kong-Response-Latency: 1
    Server: kong/3.0.1

>     {"message":"Unauthorized"

Reviewer: ```suggestion
    {"message":"Unauthorized"}
```

The `iss` field in the payload matches the value you provided in `--from-literal=key=` when creating the Kubernetes secret.

1. Copy the "Encoded" value and store it in an environment variable for both the `ADMIN_JWT` and `USER_JWT`:

    ```bash
    export ADMIN_JWT=eyJhbG...
    export USER_JWT=eyJhbG...
    ```

1. Send a request with the `Authorization` header.

    ```bash

>     curl -I -H "Authorization: Bearer ${USER_JWT}" http://kong.example/lemon --connect-to kong.example:80:$PROXY_IP

Reviewer: ```suggestion
    curl -I -H 'Host: kong.example' -H "Authorization: Bearer ${USER_JWT}" $PROXY_IP/lemon
```

```
{% endnavtab %}
{% endnavtabs %}
{% endcapture %}
{{ the_code | indent }}

1. Add consumers to groups through credentials.

    ```bash
    kubectl create secret \
      generic admin-acl \
      --from-literal=kongCredType=acl  \
      --from-literal=group=admin
    
    kubectl create secret \
      generic user-acl \
      --from-literal=kongCredType=acl  \
      --from-literal=group=user
    ```
    The results should look like this:
    ```text
    secret/admin-acl created
    secret/user-acl created
    ```

1. Associate the groups to their consumers through `credentials`.

    ```bash
    kubectl patch --type json kongconsumer admin \
      -p='[{
        "op":"add",
        "path":"/credentials/-",
        "value":"admin-acl" 
      }]'
    kubectl patch --type json kongconsumer user \
      -p='[{
        "op":"add",
        "path":"/credentials/-",
        "value":"user-acl" 
      }]'
    ```
    The results should look like this:
    ```text
    kongconsumer.configuration.konghq.com/admin patched
    kongconsumer.configuration.konghq.com/user patched
    ```

### Send authorized requests

1. Send a request as the `admin` consumer to both the URLS.

    ```bash

>     curl -sI http://kong.example/lemon -H "Authorization: Bearer ${ADMIN_JWT}" --connect-to kong.example:80:$PROXY_IP | grep HTTP

Reviewer: ```suggestion
    curl -sI $PROXY_IP/lemon -H 'Host: kong.example' -H "Authorization: Bearer ${ADMIN_JWT}" | grep HTTP
```

>     curl -sI http://kong.example/lime -H "Authorization: Bearer ${ADMIN_JWT}" --connect-to kong.example:80:$PROXY_IP | grep HTTP

Reviewer: ```suggestion
     curl -sI $PROXY_IP/lime -H 'Host: kong.example' -H "Authorization: Bearer ${ADMIN_JWT}" | grep HTTP
```

    ```
    The results should look like this:
    ```text
    HTTP/1.1 200 OK
    HTTP/1.1 200 OK
    ```

1. Send a request as the`user` consumer.

    ```bash

>     curl -sI http://kong.example/lemon -H "Authorization: Bearer ${USER_JWT}" --connect-to kong.example:80:$PROXY_IP | grep HTTP
>     curl -sI http://kong.example/lime -H "Authorization: Bearer ${USER_JWT}" --connect-to kong.example:80:$PROXY_IP | grep HTTP

Reviewer: ```suggestion
    curl -sI $PROXY_IP/lemon -H 'Host: kong.example' -H "Authorization: Bearer ${USER_JWT}" | grep HTTP
    curl -sI $PROXY_IP/lime -H 'Host: kong.example' -H "Authorization: Bearer ${USER_JWT}" | grep HTTP
```

    KjptNFAsgUjWneZFC0toI7ivKudCQWB6v/fBn/7Lycd/qT4DOaYJsE/up23qnsf+
    U8y3emZ8F+s69T5k5aEzIxs89HD7zElKNHCSlUIl+Gar0h1HTN1QkzuGwppgGrYI
    H1nkAGYMjhSJa8TvYDI/Eze4KhYwjfNEnRbuYB74LyGl5Z2imJYlOPb2bYTAtmw0
    2GNDj27LGFbfhPA4rUE8EadIsE4i4AIdl8UH3OnAYsyj2/Ubr+Z7TtTZYsRlv4MR
    FVKjO71wonAD3ssZXGz7DKleqvXedEPGXeLbz2BrDKsOGojawLA04yNY0xMT0slN
    g/KPYWmuug0oONPZhKYsj/H0Jk1xiCyJK9B6ItfbAK021NWXrAPHbgUNEbNFy8yS
    7cPgS3OgEoCMBITKVpJlxvEwuajrewBRUcgS9opz+J94dGd8bnmLRlXK6290Ni6S
    hF9M5348YA/8VIjYq7XBw9sCAwEAAQ==
    -----END PUBLIC KEY-----"
    ```

   The results should look like this:
    
    ```text
    secret/admin-jwt created
    secret/user-jwt created
    ```
   To associate the JWT Secrets with your consumers, you must add their name to the `credentials` array in the KongConsumers.

1. Assign the credentials `admin-jwt` to the `admin`.     

    ```bash
    kubectl patch --type json kongconsumer admin \
      -p='[{
        "op":"add",
        "path":"/credentials",
        "value":["admin-jwt"]
      }]'
    ```
    The results should look like this:
    ```text
    kongconsumer.configuration.konghq.com/admin patched
    ```

> 1. Assign the credentials `user-wt` to the `user`. 

Reviewer: ```suggestion
1. Assign the credentials `user-jwt` to the `user`. 
```

The `iss` field in the payload matches the value you provided in `--from-literal=key=` when creating the Kubernetes secret.

1. Copy the "Encoded" value and store it in an environment variable for both the `ADMIN_JWT` and `USER_JWT`:

    ```bash
    export ADMIN_JWT=eyJhbG...
    export USER_JWT=eyJhbG...
    ```

1. Send a request with the `Authorization` header.

    ```bash

>     curl -I -H "Authorization: Bearer ${USER_JWT}" http://kong.example/lemon --connect-to kong.example:80:$PROXY_IP

Reviewer: Why does this use `--connect-to` rather than `Host: kong.example`?

Author: I get this error: rajakavitha@198 ~ % curl -I -H "Host:kong.example:80" $PROXY_IP/lemon                                    
HTTP/1.1 401 Unauthorized
Date: Wed, 04 Oct 2023 06:09:58 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Content-Length: 26
X-Kong-Response-Latency: 3
Server: kong/3.3.1

rajakavitha@198 ~ % echo $USER_JWT
eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ1c2VyLWlzc3VlciJ9.KrdGJ50DYHiElgDULXDl3ZapcHIyILELFoGTIdYB54JTOxIgsaVo6aei2QbYucLK0GCmPNx7ED9J9SfilK_6ZsR7wHq7f7kGozXwgWf3LG5gMak1mzoHnJPi9cwzbSpV-OhJ4q4LkFUi7sGxRPlvSiP2GnxU38rxTBOonJ_y_nkZLXnTlaXanMxCdW_sccP9Y6_Px3_NtoRpVwkf11vj0Dv_9NMqK7WXEj8VPafOMp7NXFSZ0Ebo6vJ89znsA3_XdUqL0_PPZXMZ3ehPVfgz766LTJ8rf1_sQ0W6wxPapXEyiid_eo7JgAtVG4jYHspzuobkmWnJR7lLJOKlm8tVlLQo0e8nOVqV3Ks5e-kqHCfPG-5lD81u5gprG9UB04M8vVw5VN5uT5yHvGWEN2YmA9QsvFjHqTKa7lr0gTXS7lNBCaDLrm2YBSrywZMqs_QOlUtZLTEQSr1B6-cpp_b8gNxNsxzUctca3zj93QJ593qC2ifKxw8XkMab-YhV_lJexjwf1SO_AOBALTIcficI_Z5BVfLWjnEFOl2mEDffDs36fLft-d82tTSYQ76iCgy3tYEqdb3mwPsamaHIcbSYs35xoOocv1kKq81Pq-h6l3aX3yvj9OPfbfO2eS-faC2R0rd6PcjTIzD1v4f48gatnqpp_cKmu0tPfZ3c3dQLP5I
rajakavitha@198 ~ % echo $ADMIN_JWT
eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJhZG1pbi1pc3N1ZXIifQ.kiHOLThl1SggXGxNhfDuSqP70uf4V80HUoeVBt7S-YqMz0DEL6IUeafMl_LzZyLYQO0qhhNzKL10tBN0krSe3IXsS9t4tlG9kWScDT7BavDyLRAffQ-ylT-PFNk0aXNtQ-0PEN4CVu71j7t7XZcV1USAFpLlr_QD3ruvlnOA2EP_5Cw7Ub9LnGOI_TYbkmYt-OsGXfFsuofytOYeG2_CDMCSNyJC5kNyehoDrFv8JZw4MLwPbOsCbvWo6y0TKN0ydVYyHCc5eu1eqOvdnQlBmL8uBmgBm_9MAyixhbNCum3nRoevb6ySGkCZGwpvEL7vomJcLOSM1naG8OhwNguHiBHt_WkS2FrlACd11xUkFkrtoH7UMpXIgiFOBihHWfEUBfzZcIcrgmsF0rtAo64JXs-km1gwKyIfwb29XQEEWd2-KzEDa2zS7IRQXD9i1U9tGpG7q3PJZLHFGkSuKs4i5mQ8gTPr-lB7X45Yp6g-xqLSA5EvgIo78AJ-G2gJW221VZM5gOvErEmLtDjT9Kk9Cqp-L4vXpWTpAo3sJN2kaQU24O6L8ujDn2XUnNkiU7QDPOQ9TPl5lDhw0VaFtMbGkBJORc2p-3-FTQfV80X6y9RHdyrb3UOhm1rjum2DsOUaacFJPhNcVsDYKTo1bhKoqVDtMYUvNjWP6hawb5H4jyg
rajakavitha@198 ~ % curl -I -H "Authorization: Bearer ${USER_JWT}" http://kong.example/lemon --resolve kong.example:80:$PROXY_IP

curl: (49) Couldn't parse CURLOPT_RESOLVE entry 'kong.example:80:198.19.249.2:80'
rajakavitha@198 ~ % curl -I -H "Authorization: Bearer ${USER_JWT}" http://$PROXY_IP/lemon                           
HTTP/1.1 404 Not Found
Date: Wed, 04 Oct 2023 06:13:56 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Content-Length: 52
X-Kong-Response-Latency: 1
Server: kong/3.3.1

rajakavitha@198 ~ % curl -I -H "Authorization: Bearer ${USER_JWT}" http://kong.example/lemon --resolve kong.example:80:$PROXY_IP

curl: (49) Couldn't parse CURLOPT_RESOLVE entry 'kong.example:80:198.19.249.2:80'
rajakavitha@198 ~ % curl -I -H "Authorization: Bearer ${USER_JWT}" http://kong.example/lemon --connect-to kong.example:80:$PROXY_IP

HTTP/1.1 200 OK
Content-Type: text/plain; charset=utf-8
Content-Length: 137
Connection: keep-alive
Date: Wed, 04 Oct 2023 06:17:43 GMT
X-Kong-Upstream-Latency: 4
X-Kong-Proxy-Latency: 8
Via: kong/3.3.1

rajakavitha@198 ~ % echo "
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: admin-acl
plugin: acl
config:
  allow: ['admin']
" | kubectl apply -f -

kongplugin.configuration.konghq.com/admin-acl created
rajakavitha@198 ~ % echo "
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: anyone-acl
plugin: acl
config:
  allow: ['admin','user']
" | kubectl apply -f -

kongplugin.configuration.konghq.com/anyone-acl created
rajakavitha@198 ~ % kubectl annotate ingress lemon konghq.com/plugins=admin-acl
kubectl annotate ingress lime konghq.com/plugins=anyone-acl

ingress.networking.k8s.io/lemon annotated
ingress.networking.k8s.io/lime annotated
rajakavitha@198 ~ % kubectl create secret \
  generic admin-acl \
  --from-literal=kongCredType=acl  \
  --from-literal=group=admin

kubectl create secret \
  generic user-acl \
  --from-literal=kongCredType=acl  \
  --from-literal=group=user

secret/admin-acl created
secret/user-acl created
rajakavitha@198 ~ % kubectl patch --type json kongconsumer admin \
  -p='[{
    "op":"add",
    "path":"/credentials/-",
    "value":"admin-acl"
  }]'
kubectl patch --type json kongconsumer user \
  -p='[{
    "op":"add",
    "path":"/credentials/-",
    "value":"user-acl"
  }]'

kongconsumer.configuration.konghq.com/admin patched
kongconsumer.configuration.konghq.com/user patched
rajakavitha@198 ~ % curl -sI http://kong.example/lemon -H "Authorization: Bearer ${ADMIN_JWT}" --resolve kong.example:80:$PROXY_IP | grep HTTP
curl -sI http://kong.example/lime -H "Authorization: Bearer ${ADMIN_JWT}" --resolve kong.example:80:$PROXY_IP | grep HTTP

rajakavitha@198 ~ % curl -sI http://kong.example/lemon -H "Authorization: Bearer ${ADMIN_JWT}" --connect-to kong.example:80:$PROXY_IP | grep HTTP
curl -sI http://kong.example/lime -H "Authorization: Bearer ${ADMIN_JWT}" --connect-to kong.example:80:$PROXY_IP | grep HTTP

HTTP/1.1 200 OK
HTTP/1.1 200 OK
rajakavitha@198 ~ % curl -sI http://kong.example/lemon -H "Authorization: Bearer ${USER_JWT}" --connect-to kong.example:80:$PROXY_IP | grep HTTP
curl -sI http://kong.example/lime -H "Authorization: Bearer ${USER_JWT}" --connect-to kong.example:80:$PROXY_IP | grep HTTP

HTTP/1.1 403 Forbidden
HTTP/1.1 200 OK
rajakavitha@198 ~ %

Reviewer: I think this is unrelated. I fixed an issue around the `401` issue. Please update to use `Host` and if there's still a `401` I will investigate

Author: Fixed!!
